src/libData/DataStructures/CircularArray.h

```diff
@@ -73,7 +81,8 @@ template<class T> class CircularArray
             LOG_GENERAL(WARNING, "m_array is empty")
             throw;
         }
-        m_array[(int)(index % m_capacity)] = element;
+        m_index = (int)index % m_capacity;
+        m_array[m_index] = element;
         m_size++;
```

Comment:
Although this is not exposed, the size can't keep increasing. Should be `std::max(m_size++, m_capacity)`

Reply:
The usage of this variable is such that it tells the total size of the elements in the array, it would cause some internal issues if we do that.

Reply:
> would cause some internal issues if we do that

So true! I don't know if what's there right now works. Highly doubt it...  But @digiront - did you see [this](https://github.com/Zilliqa/Zilliqa/blob/f0d8d388fe907dd73b85a7b8e329cc7872da92db/src/libData/BlockChainData/DSBlockChain.cpp#L49), [this](https://github.com/Zilliqa/Zilliqa/blob/f0d8d388fe907dd73b85a7b8e329cc7872da92db/src/libData/BlockChainData/TxBlockChain.cpp#L49) and [this](https://github.com/Zilliqa/Zilliqa/blob/dc6389e473862a14da3933d60efcc6d388c54971/src/libData/BlockChainData/VCBlockChain.cpp#L43) (copies)? What about [this](https://github.com/Zilliqa/Zilliqa/blob/9f2aef617ca8ba0cfd8bd361eed63cfd268d122b/src/libServer/Server.cpp#L832) and [this](https://github.com/Zilliqa/Zilliqa/blob/9f2aef617ca8ba0cfd8bd361eed63cfd268d122b/src/libServer/Server.cpp#L939) or [this](https://github.com/Zilliqa/Zilliqa/blob/9f2aef617ca8ba0cfd8bd361eed63cfd268d122b/src/libServer/Server.cpp#L1003)?! It's so funny!! So effectively they have your 'max' suggestion - but they just have it outside the class (in the latter case) and not in the first one. Likely they use this as a circular buffer and something fancy - different - like a thing where they put known blocks and if they've seen more than expected (even if some of them are duplicates overriding the old values - a bug) something happens.

Reply:
This circular array here is not just a circular buffer. It's designed to be able to receive element with inconsistent index, thus insert at the responding position, and be able to returns the element in a desired index. At meanwhile, it will recycle the space for the latest `m_capacity` interested insertion.

Reply:
I'm sorry. This implementation you have is not correct, though, even with what you just said. :( It's better than it used to be but it still doesn't work.

Reply:
Those are simple stuff - we shouldn't talk about them, compromise etc. They can easily be 100% correct.

Reply:
> So effectively they have your 'max' suggestion - but they just have it outside the class

I guess we can all agree this should be inside the class.

> Likely they use this as a circular buffer and something fancy

That begs for bugs to appear, I think we can also agree to refactor a little bit to make the special usage obvious.